File: apps/android/app/src/main/java/com/litter/android/state/SavedServerStore.kt
package com.litter.android.state

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONArray
import org.json.JSONObject
import uniffi.codex_mobile_client.AppDiscoveredServer
import uniffi.codex_mobile_client.AppDiscoverySource
import uniffi.codex_mobile_client.SavedServerRecord

/**
 * Persistent server list stored in SharedPreferences.
 * Platform-specific — cannot live in Rust.
 */
data class SavedServer(
    val id: String,
    val name: String,
    val hostname: String,
    val port: Int,
    val agentPorts: List<Int> = emptyList(),
    val sshPort: Int? = null,
    val source: String = "manual", // local, bonjour, tailscale, lanProbe, arpScan, ssh, manual
    val hasAgentServer: Boolean = false,
    val wakeMAC: String? = null,
    val preferredConnectionMode: String? = null, // direct or ssh
    val preferredAgentPort: Int? = null,
    val sshPortForwardingEnabled: Boolean? = null, // legacy migration only
    val websocketURL: String? = null,
    val os: String? = null,
    val sshBanner: String? = null,
    val rememberedByUser: Boolean = false,
) {
    /** Stable key for deduplication across discovery cycles. */
    val deduplicationKey: String
        get() = websocketURL ?: normalizedHostKey(hostname)

    private fun normalizedHostKey(host: String): String {
        val trimmed = host.trim().trimStart('[').trimEnd(']')
        val withoutScope = if (!trimmed.contains(":")) {
            trimmed.substringBefore('%')
        } else {
            trimmed
        }
        return withoutScope.lowercase()
    }

    fun toJson(): JSONObject = JSONObject().apply {
        put("id", id)
        put("name", name)
        put("hostname", hostname)
        put("port", port)
        put("agentPorts", JSONArray(availableDirectAgentPorts))
        sshPort?.let { put("sshPort", it) }
        put("source", source)
        put("hasAgentServer", hasAgentServer)
        wakeMAC?.let { put("wakeMAC", it) }
        preferredConnectionMode?.let { put("preferredConnectionMode", it) }
        preferredAgentPort?.let { put("preferredAgentPort", it) }
        sshPortForwardingEnabled?.let { put("sshPortForwardingEnabled", it) }
        websocketURL?.let { put("websocketURL", it) }
        os?.let { put("os", it) }
        sshBanner?.let { put("sshBanner", it) }
        put("rememberedByUser", rememberedByUser)
    }

    val availableDirectAgentPorts: List<Int>
        get() {
            val ordered = buildList {
                if (hasAgentServer && port > 0) add(port)
                addAll(agentPorts.filter { it > 0 })
            }
            return ordered.distinct()
        }

    val resolvedPreferredConnectionMode: String?
        get() = when (preferredConnectionMode) {
            "direct", "directCodex" -> if (availableDirectAgentPorts.isNotEmpty() || websocketURL != null) "direct" else null
            "ssh" -> if (canConnectViaSsh) "ssh" else null
            else -> if (sshPortForwardingEnabled == true) "ssh" else null
        }

    val prefersSshConnection: Boolean
        get() = resolvedPreferredConnectionMode == "ssh"

    val canConnectViaSsh: Boolean
        get() = websocketURL == null && (
            sshPort != null ||
                source == "ssh" ||
                (!hasAgentServer && resolvedSshPort > 0) ||
                preferredConnectionMode == "ssh" ||
                sshPortForwardingEnabled == true
        )

    val resolvedSshPort: Int
        get() = sshPort ?: port.takeIf { !hasAgentServer && it > 0 } ?: 22

    val resolvedPreferredAgentPort: Int?
        get() = when {
            resolvedPreferredConnectionMode != "direct" -> null
            preferredAgentPort != null && availableDirectAgentPorts.contains(preferredAgentPort) -> preferredAgentPort
            else -> null
        }

    val requiresConnectionChoice: Boolean
        get() = websocketURL == null &&
            resolvedPreferredConnectionMode == null &&
            (
                availableDirectAgentPorts.size > 1 ||
                    (availableDirectAgentPorts.isNotEmpty() && canConnectViaSsh)
            )

    val directAgentPort: Int?
        get() = when {
            websocketURL != null -> null
            prefersSshConnection -> null
            resolvedPreferredAgentPort != null -> resolvedPreferredAgentPort
            requiresConnectionChoice -> null
            availableDirectAgentPorts.isNotEmpty() -> availableDirectAgentPorts.first()
            else -> null
        }

    fun withPreferredConnection(mode: String?, agentPortValue: Int? = null): SavedServer =
        copy(
            port = when (mode) {
                "direct" -> agentPortValue ?: directAgentPort ?: availableDirectAgentPorts.firstOrNull() ?: port
                "ssh" -> resolvedSshPort
                else -> port
            },
            agentPorts = availableDirectAgentPorts,
            sshPort = sshPort ?: if (canConnectViaSsh) resolvedSshPort else null,
            preferredConnectionMode = mode,
            preferredAgentPort = if (mode == "direct") {
                agentPortValue ?: directAgentPort ?: availableDirectAgentPorts.firstOrNull()
            } else {
                null
            },
            sshPortForwardingEnabled = null,
        )

    fun normalizedForPersistence(): SavedServer = withPreferredConnection(
        mode = resolvedPreferredConnectionMode,
        agentPortValue = resolvedPreferredAgentPort ?: availableDirectAgentPorts.firstOrNull(),
    )

    companion object {
        fun normalizeWakeMac(raw: String?): String? {
            val compact = raw
                ?.trim()
                ?.replace(":", "")
                ?.replace("-", "")
                ?.lowercase()
                ?: return null
            if (compact.length != 12 || compact.any { !it.isDigit() && it !in 'a'..'f' }) {
                return null
            }
            return buildString {
                compact.chunked(2).forEachIndexed { index, chunk ->
                    if (index > 0) append(':')
                    append(chunk)
                }
            }
        }

        fun fromJson(obj: JSONObject): SavedServer = SavedServer(
            id = obj.getString("id"),
            name = obj.optString("name", ""),
            hostname = obj.optString("hostname", ""),
            port = obj.optInt("port", 0),
            agentPorts = buildList {
                val ports = obj.optJSONArray("agentPorts") ?: obj.optJSONArray("codexPorts")
                if (ports != null) {
                    for (index in 0 until ports.length()) {
                        add(ports.optInt(index))
                    }
                }
            },
            sshPort = if (obj.has("sshPort")) obj.getInt("sshPort") else null,
            source = obj.optString("source", "manual"),
            hasAgentServer = obj.optBoolean("hasAgentServer", obj.optBoolean("hasCodexServer", false)),
            wakeMAC = if (obj.has("wakeMAC")) obj.getString("wakeMAC") else null,
            preferredConnectionMode = obj.optString("preferredConnectionMode").ifBlank { null }
                ?.let { mode ->
                    // Migrate legacy "directCodex" to "direct"
                    if (mode == "directCodex") "direct" else mode
                },
            preferredAgentPort = if (obj.has("preferredAgentPort")) {
                obj.getInt("preferredAgentPort")
            } else if (obj.has("preferredCodexPort")) {
                obj.getInt("preferredCodexPort")
            } else {
                null
            },
            sshPortForwardingEnabled = if (obj.has("sshPortForwardingEnabled")) {
                obj.optBoolean("sshPortForwardingEnabled")
            } else {
                null
            },
            websocketURL = if (obj.has("websocketURL")) obj.getString("websocketURL") else null,
            os = if (obj.has("os")) obj.getString("os") else null,
            sshBanner = if (obj.has("sshBanner")) obj.getString("sshBanner") else null,
            rememberedByUser = if (obj.has("rememberedByUser")) {
                obj.optBoolean("rememberedByUser")
            } else {
                true
            },
        )

        fun from(server: AppDiscoveredServer): SavedServer = SavedServer(
            id = server.id,
            name = server.displayName,
            hostname = server.host,
            port = server.agentPort?.toInt() ?: server.port.toInt(),
            agentPorts = server.agentPorts.map { it.toInt() },
            sshPort = server.sshPort?.toInt(),
            source = when (server.source) {
                AppDiscoverySource.BONJOUR -> "bonjour"
                AppDiscoverySource.TAILSCALE -> "tailscale"
                AppDiscoverySource.LAN_PROBE -> "lanProbe"
                AppDiscoverySource.ARP_SCAN -> "arpScan"
                AppDiscoverySource.MANUAL -> "manual"
                AppDiscoverySource.LOCAL -> "local"
            },
            hasAgentServer = server.agentPort != null || server.agentPorts.isNotEmpty(),
            os = if (server.sshBanner != null) server.os else server.os,
            sshBanner = server.sshBanner,
        )
    }
}

fun SavedServer.toRecord() = SavedServerRecord(
    id = id,
    name = name,
    hostname = hostname,
    port = port.toUShort(),
    agentPorts = agentPorts.map { it.toUShort() },
    sshPort = sshPort?.toUShort(),
    source = source,
    hasAgentServer = hasAgentServer,
    wakeMac = wakeMAC,
    preferredConnectionMode = preferredConnectionMode,
    preferredAgentPort = preferredAgentPort?.toUShort(),
    sshPortForwardingEnabled = sshPortForwardingEnabled,
    websocketUrl = websocketURL,
    rememberedByUser = rememberedByUser,
)

object SavedServerStore {
    private const val PREFS_NAME = "codex_saved_servers_prefs"
    private const val KEY = "codex_saved_servers"

    private fun prefs(context: Context): SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    fun load(context: Context): List<SavedServer> {
        val json = prefs(context).getString(KEY, null) ?: return emptyList()
        return try {
            val array = JSONArray(json)
            val decoded = (0 until array.length()).map { SavedServer.fromJson(array.getJSONObject(it)) }
            val migrated = decoded.map { it.normalizedForPersistence() }
            if (decoded != migrated) {
                save(context, migrated)
            }
            migrated
        } catch (_: Exception) {
            emptyList()
        }
    }

    fun save(context: Context, servers: List<SavedServer>) {
        val array = JSONArray()
        servers.forEach { array.put(it.toJson()) }
        prefs(context).edit().putString(KEY, array.toString()).apply()
    }

    fun upsert(context: Context, server: SavedServer) {
        val existing = load(context).toMutableList()
        val prior = existing.firstOrNull { it.id == server.id || it.deduplicationKey == server.deduplicationKey }
        existing.removeAll { it.id == server.id || it.deduplicationKey == server.deduplicationKey }
        existing.add(server.copy(rememberedByUser = prior?.rememberedByUser ?: server.rememberedByUser))
        save(context, existing)
    }

    fun remember(context: Context, server: SavedServer) {
        val existing = load(context).toMutableList()
        existing.removeAll { it.id == server.id || it.deduplicationKey == server.deduplicationKey }
        existing.add(server.copy(rememberedByUser = true))
        save(context, existing)
    }

    fun remembered(context: Context): List<SavedServer> =
        load(context).filter { it.rememberedByUser }

    fun remove(context: Context, serverId: String) {
        val existing = load(context).toMutableList()
        existing.removeAll { it.id == serverId }
        save(context, existing)
    }

    fun rename(context: Context, serverId: String, newName: String) {
        val trimmed = newName.trim()
        if (trimmed.isEmpty()) return

        val existing = load(context)
        val renamed = existing.map { server ->
            if (server.id == serverId) server.copy(name = trimmed) else server
        }
        if (renamed != existing) {
            save(context, renamed)
        }
    }

    fun updateWakeMac(context: Context, serverId: String, host: String, wakeMac: String?) {
        val normalizedWakeMac = SavedServer.normalizeWakeMac(wakeMac) ?: return
        val existing = load(context)
        val updated = existing.map { server ->
            if (server.id == serverId || normalizedHostKey(server.hostname) == normalizedHostKey(host)) {
                if (server.wakeMAC != normalizedWakeMac) server.copy(wakeMAC = normalizedWakeMac) else server
            } else {
                server
            }
        }
        if (updated != existing) {
            save(context, updated)
        }
    }

    private fun normalizedHostKey(host: String): String {
        val trimmed = host.trim().trimStart('[').trimEnd(']').replace("%25", "%")
        val withoutScope = if (!trimmed.contains(":")) {
            trimmed.substringBefore('%')
        } else {
            trimmed
        }
        return withoutScope.lowercase()
    }
}

File: apps/ios/Sources/Litter/Models/SavedServerStore.swift
import Foundation
import UIKit

@MainActor
enum SavedServerStore {
    private static let savedServersKey = "codex_saved_servers"

    static func save(_ servers: [SavedServer]) {
        guard let data = try? JSONEncoder().encode(servers) else { return }
        UserDefaults.standard.set(data, forKey: savedServersKey)
    }

    static func load() -> [SavedServer] {
        guard let data = UserDefaults.standard.data(forKey: savedServersKey) else { return [] }
        let decoded = (try? JSONDecoder().decode([SavedServer].self, from: data)) ?? []
        let migrated = decoded.map { saved -> SavedServer in
            let server = saved.toDiscoveredServer()
            let restored = SavedServer.from(server, rememberedByUser: saved.rememberedByUser)
            if shouldReplaceLegacyLocalPlaceholder(restored) {
                return SavedServer(
                    id: restored.id,
                    name: UIDevice.current.name,
                    hostname: restored.hostname,
                    port: restored.port,
                    agentPorts: restored.agentPorts,
                    sshPort: restored.sshPort,
                    source: restored.source,
                    hasAgentServer: restored.hasAgentServer,
                    wakeMAC: restored.wakeMAC,
                    preferredConnectionMode: restored.preferredConnectionMode,
                    preferredAgentPort: restored.preferredAgentPort,
                    sshPortForwardingEnabled: restored.sshPortForwardingEnabled,
                    websocketURL: restored.websocketURL,
                    rememberedByUser: restored.rememberedByUser
                )
            }
            return restored
        }
        if migrated != decoded {
            save(migrated)
        }
        return migrated
    }

    static func upsert(_ server: DiscoveredServer) {
        var saved = load()
        let existing = existingMatch(for: server, in: saved)
        saved.removeAll { entry in matches(server, entry) }
        saved.append(
            SavedServer.from(
                server,
                rememberedByUser: existing?.rememberedByUser ?? false
            )
        )
        save(saved)
    }

    static func remember(_ server: DiscoveredServer) {
        var saved = load()
        saved.removeAll { entry in matches(server, entry) }
        saved.append(SavedServer.from(server, rememberedByUser: true))
        save(saved)
    }

    static func rememberedServers() -> [SavedServer] {
        load().filter(\.rememberedByUser)
    }

    static func reconnectRecords(
        localDisplayName: String,
        rememberedOnly: Bool = false
    ) -> [SavedServerRecord] {
        let saved = rememberedOnly ? rememberedServers() : load()
        var records = saved.map { $0.toRecord() }
        if records.contains(where: { $0.id == "local" || $0.source == ServerSource.local.rawValue }) == false {
            records.append(
                SavedServerRecord(
                    id: "local",
                    name: localDisplayName,
                    hostname: "127.0.0.1",
                    port: 0,
                    agentPorts: [],
                    sshPort: nil,
                    source: ServerSource.local.rawValue,
                    hasAgentServer: true,
                    wakeMac: nil,
                    preferredConnectionMode: nil,
                    preferredAgentPort: nil,
                    sshPortForwardingEnabled: nil,
                    websocketUrl: nil,
                    rememberedByUser: true
                )
            )
        }
        return records
    }

    static func remove(serverId: String) {
        var saved = load()
        saved.removeAll { $0.id == serverId }
        save(saved)
    }

    static func rename(serverId: String, newName: String) {
        var saved = load()
        guard let index = saved.firstIndex(where: { $0.id == serverId }) else { return }
        let old = saved[index]
        saved[index] = SavedServer(
            id: old.id,
            name: newName,
            hostname: old.hostname,
            port: old.port,
            agentPorts: old.agentPorts,
            sshPort: old.sshPort,
            source: old.source,
            hasAgentServer: old.hasAgentServer,
            wakeMAC: old.wakeMAC,
            preferredConnectionMode: old.preferredConnectionMode,
            preferredAgentPort: old.preferredAgentPort,
            sshPortForwardingEnabled: old.sshPortForwardingEnabled,
            websocketURL: old.websocketURL,
            rememberedByUser: old.rememberedByUser
        )
        save(saved)
    }

    static func updateWakeMAC(serverId: String, host: String, wakeMAC: String?) {
        guard let normalizedWakeMAC = DiscoveredServer.normalizeWakeMAC(wakeMAC) else { return }

        var saved = load()
        guard let index = saved.firstIndex(where: { entry in
            entry.id == serverId || normalizedHost(entry.hostname) == normalizedHost(host)
        }) else {
            return
        }

        let existing = saved[index]
        guard existing.wakeMAC != normalizedWakeMAC else { return }

        saved[index] = SavedServer(
            id: existing.id,
            name: existing.name,
            hostname: existing.hostname,
            port: existing.port,
            agentPorts: existing.agentPorts,
            sshPort: existing.sshPort,
            source: existing.source,
            hasAgentServer: existing.hasAgentServer,
            wakeMAC: normalizedWakeMAC,
            preferredConnectionMode: existing.preferredConnectionMode,
            preferredAgentPort: existing.preferredAgentPort,
            sshPortForwardingEnabled: existing.sshPortForwardingEnabled,
            websocketURL: existing.websocketURL,
            rememberedByUser: existing.rememberedByUser
        )
        save(saved)
    }

    private static func existingMatch(for server: DiscoveredServer, in saved: [SavedServer]) -> SavedServer? {
        saved.first { matches(server, $0) }
    }

    private static func matches(_ server: DiscoveredServer, _ savedServer: SavedServer) -> Bool {
        savedServer.id == server.id || savedServer.toDiscoveredServer().deduplicationKey == server.deduplicationKey
    }

    private static func normalizedHost(_ host: String) -> String {
        var normalized = host
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .trimmingCharacters(in: CharacterSet(charactersIn: "[]"))
            .replacingOccurrences(of: "%25", with: "%")

        if !normalized.contains(":"), let scopeIndex = normalized.firstIndex(of: "%") {
            normalized = String(normalized[..<scopeIndex])
        }

        return normalized.lowercased()
    }

    private static func shouldReplaceLegacyLocalPlaceholder(_ server: SavedServer) -> Bool {
        server.source == .local
            && server.name.trimmingCharacters(in: .whitespacesAndNewlines) == "This Device"
    }
}
